### PR TITLE
Add begin_tx and end_tx specs

### PR DESCRIPTION
**Pull request by rrtoledo**
Adding markdown spec for begin_tx & end_tx gadgets

### CODE REVIEW

**Comment by rrtoledo** [comment]
There are notable differences between the python and the rs code (e.g. InvalidTx is used in python, precompiles are more handled in rust...). I'll set up a call to talk about these.

**Comment by rrtoledo** [comment]
We are handling invalidtx differently in the specs than in rust (where we do not handle them there in begin/end_tx). Should we update all specs to sync? I think Taiko was working on the rust's InvalidTx.

**Comment by rrtoledo** [comment]
There are differences between src/* and build/lib/*, @ChihChengLiang pointed out that build/ is ignored, would it be worth deleting the folder?

**Comment by rrtoledo** [comment]
Precompiles are handled in rust but not in the specs, should we open an issue about it?

**Comment by KimiWu123** [comment]
> We are handling invalidtx differently in the specs than in rust (where we do not handle them there in begin/end_tx). Should we update all specs to sync? I think Taiko was working on the rust's InvalidTx.

We have invalid tx support in rust (https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1677). The implementation is more comprehensive in rust. So, yes, we should update our specs to sync our circuit.

> Precompiles are handled in rust but not in the specs, should we open an issue about it?

We have precompiles in the specs. The precompile gadgets in the specs should be more than in the circuit.

File: specs/virtual_state/end_tx.md
Comment (KimiWu123) [inline specs/virtual_state/end_tx.md:68]: Should be `end_tx.py`.

File: specs/virtual_state/end_tx.md
Comment (KimiWu123) [inline specs/virtual_state/end_tx.md:9]: `CREATE`

**Review by KimiWu123 — APPROVED** [review]
LGTM. Great work!!  Left some minor suggestions.

File: specs/virtual_state/begin_tx.md
Comment (ChihChengLiang) [reply specs/virtual_state/begin_tx.md:22]: I've been thinking we shouldn't spell out all the arithmetization details in the spec. Unless we intentionally deploy some optimization tricks, otherwise we can abstract out the arithmetization details most of the time.
It would already be super helpful to mention just the three conditions that must be satisfied.
- sufficient balance: sender balance > value + fee
- sufficient gas: gas limit > gas used
- valid nonce: sender nonce = tx nonce

**Review by ChihChengLiang — COMMENTED** [review]
Generally looks great! I added some comments

File: specs/virtual_state/begin_tx.md
Comment (rrtoledo) [reply specs/virtual_state/begin_tx.md:24]: "calldata" that's a typo!

**Comment by rrtoledo** [comment]
> > We are handling invalidtx differently in the specs than in rust (where we do not handle them there in begin/end_tx). Should we update all specs to sync? I think Taiko was working on the rust's InvalidTx.
> 
> We have invalid tx support in rust ([privacy-scaling-explorations/zkevm-circuits#1677](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1677)). The implementation is more comprehensive in rust. So, yes, we should update our specs to sync our circuit.
> 
> > Precompiles are handled in rust but not in the specs, should we open an issue about it?
> 
> We have precompiles in the specs. The precompile gadgets in the specs should be more than in the circuit.

I'll add the precompiles to the access list then!

**Review by ChihChengLiang — APPROVED** [review]
LGTM after @KimiWu123's feedback is addressed.

File: specs/virtual_state/begin_tx.md
Comment (rrtoledo) [reply specs/virtual_state/begin_tx.md:32]: I'll remove this. That was just a python error because precompiles are not handled well there.

**Comment by rrtoledo** [comment]
Updates:
- Removing refs to python code and adding some to rust code (gadgets)
- Removing Invalidtx refs (does not exist in rust code)
- Adding precompiles to access list
- Updating end_tx description!
- Removing the arithmetization details 
- Adding more verbose constraints, formatting Create to CREATE
- Correcting lookups

File: specs/virtual_state/begin_tx.md
Comment (rrtoledo) [reply specs/virtual_state/begin_tx.md]: Updated

**Comment by rrtoledo** [comment]
@KimiWu123 @ChihChengLiang Sorry for bothering you again but can you have another look as the PR changed a fair bit!

File: specs/virtual_state/end_tx.md
Comment (KimiWu123) [inline specs/virtual_state/end_tx.md:19]: From the implementation, rw_counter should be "8 + transfer.rw `-` is_first_tx".
It's the same in 4.1 below.

File: specs/virtual_state/end_tx.md
Comment (KimiWu123) [inline specs/virtual_state/end_tx.md]: TBH, I like this paragraph (L7 - L13). It explain this gedget in in high level. But it's just my personal preference.

**Review by KimiWu123 — APPROVED** [review]
Left some comment and it looks much better now! Thanks for the revision!!

File: specs/virtual_state/begin_tx.md
Comment (ChihChengLiang) [inline specs/virtual_state/begin_tx.md:48]: Nitpick: The name "calls" might collide with EVM "Message call", as the action of the "CALL" opcode performs.

What we're counting here is the number of the read-write access to the [read-write table](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md). In the language of now so called "memory consistency check", the Read-write table is our RAM and the read-write counter is the timestamp.

```suggestion
In all cases, we bump read-write-counter by 9 + #transfer:
```

**Review by ChihChengLiang — APPROVED** [review]
LGTM modulo Kimi's feedback. Thanks for adding the markdown spec.
I added some nitpicks.

File: specs/virtual_state/end_tx.md
Comment (rrtoledo) [reply specs/virtual_state/end_tx.md:19]: I have issues with this particular lookup it seems 🤦

File: specs/virtual_state/begin_tx.md
Comment (rrtoledo) [reply specs/virtual_state/begin_tx.md:67]: Done!

File: specs/virtual_state/end_tx.md
Comment (rrtoledo) [reply specs/virtual_state/end_tx.md]: I copy pasted being_tx as end_tx and then updated each section but forgot to update this one, hence the text is incorrect here.

That's why I deleted this paragraph (because this is from begin_tx!) and replaced it with:
`The 'end_tx' gadget handles refund and coinbase tip as well as cumulative gas within a block. 
`

File: specs/virtual_state/end_tx.md
Comment (rrtoledo) [reply specs/virtual_state/end_tx.md:19]: (so it is 9 + transfer.rw - is_first_tx)

Comment (rrtoledo) [reply specs/virtual_state/end_tx.md:19]: (I checked thrice)

File: specs/virtual_state/begin_tx.md
Comment (rrtoledo) [reply specs/virtual_state/begin_tx.md]: Updated in both files for all cases

File: specs/virtual_state/end_tx.md
Comment (rrtoledo) [reply specs/virtual_state/end_tx.md]: I wish I could write more but, except for the transition, that's pretty much what the gadget does

**Pull request merged**
